lib/src/models/transaction_uploader.dart

```diff
@@ -25,18 +22,21 @@ const FATAL_CHUNK_UPLOAD_ERRORS = [
   'invalid_proof'
 ];
 
-class TransactionUploader {
-  int _chunkIndex = 0;
-  bool _txPosted = false;
-  int _lastRequestTimeEnd = 0;
-  int _totalErrors = 0;
-
-  int lastResponseStatus = 0;
-  String lastResponseError = '';
+const maxConcurrentChunkUploadCount = 128;
 
+class TransactionUploader {
   final Transaction _transaction;
   final ArweaveApi _api;
-  final Random _random = Random();
+
+  bool get isComplete => _txPosted && uploadedChunks >= totalChunks;
+  int get totalChunks => _transaction.chunks!.chunks.length;
+  int get uploadedChunks => _uploadedChunks;
+
+  /// The progress of the current upload ranging from 0 to 1.
+  double get progress => uploadedChunks / totalChunks;
```

Comment:
Is `totalChunks` always greater than zero?

Reply:
Good catch! It isn't as AR only transactions do not contain data chunks to be uploaded which meant that this would've previously failed for those transactions.

I have updated the code to handle this case and added a test for this. As it wouldn't make sense for `TransactionUploader.progress` to equal 1 before the header is even uploaded for AR only transactions I had to update progress calculation to account for the header. This does mean that `dataOnly` uploads will start with progress > 0.